### DISCUSSION TITLE
ActiveObject-like object patch

### DISCUSSION THREAD

**Discussion by IlyaSemenov** [romeerez/orchid-orm#409] [Ideas]
Relatively often I find myself using this (simplified) flow:

```ts
const order = await db.order.find(orderId).select("...", "paidTime")
if (!order.paidTime) {
  await db.order.find(orderId).update({ paidTime: new Date() })
  console.log(`Order was paid at: ${order.paidTime}`) // prints undefined
}
```

This is of course expected and I'm not complaining. However, this is still:
1) a gotcha, especially for newcomers
2) needs boilerplate for proper handling, such as:

```ts
await db.order
  .find(orderId)
  .update({ paidTime: new Date() })
  .afterUpdate(["paidTime"], ([data]) => {
    Object.assign(order, data satisfies Partial<typeof order>)
  })
```

My idea is, can we somehow simplify this on ORM level? Such as:

```ts
await db.order.find(orderId).update({ paidTime: new Date() }).patch(order)
// return all updated keys and set them in the provided DAO
```

Of course there are many questions, such as how to handle multiple objects, what to do if object has no respective keys (or what if they've been selected under aliases), etc. etc. — but I just wanted to start a discussion here.

**Edit:** see #410 as well.

**Comment by romeerez** [comment]
Your `patch` looks pretty neat. But, I'd like to avoid mutations, in the sake of smarter types.

I'd like to make it so that you can have a new variable `updatedOrder` with a proper type, so TS can see that this `updatedOrder` for sure has `paidTime`, while in the `order` it is optional.

I'm not sure yet how to design this better.

Thinking of:

```ts
// data has type { paidTime: Date }, and this should override { paidTime?: Date } of the order

const updatedOrder = await db.order.find(id).update(data).assign(order)
```

If you select some fields, it will populate selected fields + updated fields into the object:

```ts
// resulting type will have both foo and bar
const updatedOrder = await db.order.find(id).update({ foo: 1 }).select('bar').assign(order)
```

If it's `findOptional`, result can be undefined.

It's not `patch` because I'd like to avoid mutations. But `assign` is also a bad name, because `Object.assign` does mutate the object, and here I'd like to not touch the original `order`. `merge`? `merge` is already taken for merging queries. Need to research on naming.

Though, `assign` is a fine name. In a sense that it's assigning from `order` to `updatedOrder`, but not the other way around.

Seems to be fine so far. What do you think about my non-mutative idea over yours mutative one?

**Reply by IlyaSemenov** [reply]
I'm not fully following why does it need the original `order` in the `.assign(order)` above? How does it use it? Looks up non-empty keys and preselects them?

Honestly I'm not convinced the whole original idea is feasible it all. Back in the time I just wanted to share my experience and thoughts but the use case could be really not worth pursuing. What troubles me now is that the new method apparently relies on the original field names/keys but the query builder makes it very easy to add aliases and the proposed shortcut couldn't have any idea about the aliased names. Unless I got it wrong somehow.

**Reply by romeerez** [reply]
I thought that `assign` could be a helper for:

```ts
const data = { paidTime: new Date() }

await db.order.find(id).update(data) // update doesn't select anything by default

const updatedOrder = { ...order, ...data }
updatedOrder.productId // from order object
updatedOrder.paidTime // is definitely set
```

```ts
const updatedOrder = await db.order.find(id).update({ paidTime: new Date() }).assign(order)
updatedOrder.productId // from order object
updatedOrder.paidTime // is definitely set, same as above, still nothing is returned from the db
```

If some fields were selected, the data for update could be merged with selected data:

```ts
const updatedOrder = await db.order.find(id).update({ paidTime: new Date() }).select('updatedAt').assign(order)
updatedOrder.productId // from order object
updatedOrder.paidTime // from update data
updatedOrder.updatedAt // loaded from db
```

So the same as yours `patch`, but without patching that order object:

```ts
await db.order.find(orderId).update({ paidTime: new Date() }).patch(order)
```

> but the use case could be really not worth pursuing.

that's good, there are plenty of things to add and improve.

**Reply by IlyaSemenov** [reply]
Right, understood. Makes sense.

I see some rigidness now in the sense that `paidTime` could have been originally selected as `aliasedPaidTime` but there is no way to provide aliases for the updating fields in the proposed API. Not saying it's a big deal, just thoughts.

**Reply by IlyaSemenov** [reply]
Oh, I actually read your message above fully and discovered the _"If some fields were selected, the data for update could be merged with selected data"_ proposal. That allows to account for aliased fields, but, how do we then prevent non-aliased updated `paidTime` being injected regardless what?

I think if there's at least one select, the fields in update should be disregarded. (Or even always disregard all fields in update and only assign those that are explicitly select'ed).

**Reply by romeerez** [reply]
> how do we then prevent non-aliased updated paidTime being injected regardless what?

You're updating `paidTime`, and while such `assign` isn't super intuitive, still it's not a big surprise if you have `paidTime` in the output.

Maybe in your case it's not desired. In other cases it's in contrary expected. So it's clear by now that such `assign` (or `patch`) functionality is saving a bit of LOCs by taking from the code obviousness.

> if there's at least one select

hidden rules

> Or even always disregard all fields in update and only assign those that are explicitly select'ed).

then there is no point in `assign`, because the fields are selected anyway

**Reply by IlyaSemenov** [reply]
> then there is no point in assign, because the fields are selected anyway

Not quite. Consider this:

```ts
const order = await db.order.find(id).select("id", "amount")
const updatedOrder = await db.order.find(id)
  .select("status", { time: "paidTime" })
  .update({ paidTime: new Date() })
  .assign(order)
```

I would expect `updatedOrder` to have id, amount, status and time (but not paidTime!)

**Reply by romeerez** [reply]
I agree by looking at your example that having both datas mixes is counter-intuitive.

In any case such `assign` addition won't be intuitive, and such JS destructuring is easier to follow:

```ts
const data = { paidTime: new Date() }

const updatedOrder = {
  ...order,
  ...data, // optionally include data if needed
  ...(
    await db.order.find(id)
      .select("status", { time: "paidTime" })
      .update(data)
  ),
}
```
